src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -91,10 +105,155 @@ trait ElectionCandidate {
   def startLeadership(): Unit
 }
 
-/** Local leadership events. They are not delivered via the event endpoints. */
-sealed trait LocalLeadershipEvent
+class ElectionServiceImpl(
+    eventStream: EventStream,
+    hostPort: String,
+    leaderEventsSource: Source[LeadershipState, Cancellable],
+    crashStrategy: CrashStrategy,
+    electionEC: ExecutionContext
+)(implicit system: ActorSystem) extends ElectionService with StrictLogging {
+
+  import ElectionService._
+  @volatile private[this] var lastState: LeadershipState = LeadershipState.Standby(None)
+  implicit private lazy val materializer = ActorMaterializer()
+  var leaderSubscription: Option[Cancellable] = None
+
+  def subscribe(subscriber: ActorRef): Unit = {
+    eventStream.subscribe(subscriber, classOf[LeadershipTransition])
+    val currentState = if (isLeader) LeadershipTransition.ElectedAsLeader else LeadershipTransition.Standby
+    subscriber ! currentState
+  }
+
+  def unsubscribe(subscriber: ActorRef): Unit = {
+    eventStream.unsubscribe(subscriber, classOf[LeadershipTransition])
+  }
+
+  override def isLeader: Boolean =
+    lastState == LeadershipState.ElectedAsLeader
 
-object LocalLeadershipEvent {
-  case object ElectedAsLeader extends LocalLeadershipEvent
-  case object Standby extends LocalLeadershipEvent
+  override def localHostPort: String = hostPort
+
+  override def leaderHostPort: Option[String] = lastState match {
+    case LeadershipState.ElectedAsLeader =>
+      Some(hostPort)
+    case LeadershipState.Standby(currentLeader) =>
+      currentLeader
+  }
+
+  /**
+    * Releases leadership
+    *
+    * Has no effect if called before offerLeadership
+    */
+  override def abdicateLeadership(): Unit = {
+    leaderSubscription.foreach(_.cancel())
+  }
+
+  private def initializeStream(leadershipTransitionsFlow: Flow[LeadershipState, LeadershipTransition, NotUsed]) = {
```

Comment:
I think this method need some docs :).

Reply:
But it's private! :)

Okay.

✅ 

This work?

```

  /**
    * Construct the stream topology and run it.
    *
    * Topology looks like this:
    *
    * +----------------------------+    +-----------+    +----------------------------+
    * |    leaderEventsSource      | -> | broadcast | -> |   localEventListenerSink   |
    * | (IE CuratorElectionStream) |    +-----------+    | (keep track of last state) |
    * |                            |          |          +----------------------------+
    * |     [LeadershipState]      |          |
    * +----------------------------+          |
    *                                         V
    *                             +--------------------------------------+
    *                             |      leadershipTransitionFlow        |
    *                             | (Emits LeadershipTransition when we  |
    *                             | lost or gain leadership, but calling |
    *                             | marathon initialization logic before |
    *                             | emitting ObtainedLeadership          |
    *                             |                                      |
    *                             |      [LeadershipTransition]          |
    *                             +--------------------------------------+
    *                                              |
    *                                              V
    *          +------------------------+    +-----------+    +------------------------+
    *          |   localTransitionSink  | <- | broadcast | -> |    metricsSink         |
    *          | Monitors for lost      |    +-----------+    | Watches transitions,   |
    *          | leadership; crashes if |                     | records how long we've |
    *          | that happens.          |                     | had leadership         |
    *          +------------------------+                     +------------------------+
    *
    * If any component throws an exception, or finishes, then the _whole stream_ shuts down.
    *
    * When the stream ends (exception or not), we report accordingly and crash.
    */
```

Reply:
😮 
Karsten: I think we need some docs.
Tim: How many pages.

Overall, I like to document code also for myself and other engineers. Ideally methods are so expressive that they need no docs. However, for complex systems it's super hard to achieve this.